File: 백준/Gold/2493. 탑/탑.py
import sys
input = sys.stdin.readline

N = int(input())
idx = 0
lst = list(map(int, input().split()))
arr = len(lst)
real = []
stack = []

while idx != arr:
    a = lst[idx]

    if stack:
        while stack:
            if stack[-1][0] < a:
                stack.pop()
                if not stack:
                    real.append(0)

            elif stack[-1][0] > a:
                real.append(stack[-1][1] + 1)
                break
            else:
                real.append(stack[-1][1] + 1)
                stack.pop()

        stack.append((a, idx))
    else:
        real.append(0)
        stack.append((a, idx))

    idx += 1

print(*real)

File: 백준/Gold/2493. 탑/탑.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static class Node {
        int high;
        int idx;

        public Node(int high, int idx) {
            this.high = high;
            this.idx = idx;
        }
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        StringBuilder sb = new StringBuilder();
        Stack<Node> stack = new Stack<>();
        int cnt = 0;

        while (cnt != n) {
            int now = arr[cnt];

            if (!stack.isEmpty()) {
                while (!stack.isEmpty()) {
                    if (stack.peek().high < now) {
                        stack.pop();
                        if (stack.isEmpty()) {
                            sb.append(0).append(" ");
                        }
                    } else {
                        sb.append(stack.peek().idx + 1).append(" ");
                        break;
                    }
                }
                stack.push(new Node(now, cnt));
            } else {
                sb.append(0).append(" ");
                stack.push(new Node(now, cnt));
            }

            cnt++;
        }
        System.out.println(sb);
    }
}